pkg/query/v2/logical/iter.go

```diff
@@ -0,0 +1,104 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package logical
+
+import (
+	"sort"
+
+	"github.com/apache/skywalking-banyandb/banyand/tsdb"
+)
+
+var _ ItemIterator = (*itemIter)(nil)
+
+type ItemIterator interface {
+	HasNext() bool
+	Next() tsdb.Item
+}
+
+// container contains both iter and its current item
+type container struct {
+	item tsdb.Item
+	iter tsdb.Iterator
+}
+
+type itemIter struct {
+	// c is the comparator to sort items
+	c comparator
+	// iters is the list of initial Iterator
+	iters []tsdb.Iterator
+	// deq is the deque of the container
+	// 1. When we push a new container, we can normally append it to the tail of the deq,
+	//    and then sort the whole slice
+	// 2. When we pop a new container, we can just pop out the first element in the deq.
+	//    The rest of the slice is still sorted.
+	deq []*container
```

Comment:
How about opts to `heap` package in std lib?

Reply:
Done